doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
+
+Otherwise, the value of the `info` attribute is a JSON object describing the
+attributes of the file.  Any attribute which cannot be read (eg: due to
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
+ * `mode`: an integer representing the "file permission" bits.  As per JSON,
+   this is transmitted as a decimal value, but it is meant to be interpreted in
+   octal, in the usual way.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
+ * `size`: an integer, the (apparent) size of the file (`st_size`)
+ * `modified`: a float, the mtime of the file (`st_mtim`)
+ * `tag`: a value type: the current 'transaction tag' of the file, with the
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
+ * `entries`: for directories: the contents of the directory (see below)
+ * `content`: for regular text files, the content (if less than 4096 bytes)
+ * `target`: for symbolic links: the target of the link (a string)
+
+The `entries` attribute is only reported on directories, and only if the
+contents of the directory can be read.  It is an object where each key is the
+name of a file in the directory, and each value is an object describing that
+file, using the same attributes as above.  `entries` is not reported on entries
+(ie: no recursive listing).
+
+The attributes to be reported can be controlled by setting the `attrs` option.
+
+If the provided `path` is a symbolic link, it is followed, unless
+`follow: false` is specified.  If symbolic links are present in the directory,
+they are always reported directly (ie: not followed).  `follow: false` is not
+supported in combination with `watch: true` but you probably don't need it:
+watch mode doesn't only watch the final path component, but all components
+leading up to it, including any symlink that have been followed.  That means
+that you can watch `/etc/localtime` and find out about if the link changed, or
+if the contents of the file itself changed, or was deleted or replaced, or also
+if `/usr`, `/etc`, or any other directories or intermediate links were changed,
+deleted, or moved out of the way.
+
+(TODO: providing `watch: true, follow: false` would be possible with an
+addition to the `systemd_ctypes` `PathWatch` API).
+
+The life-cycle of the channel works like this:
+
+ - When opening the channel, if a file at the named `path` cannot be opened and
+   `close-on-error` is true, opening fails (`close` with `problem`).
+
+ - Otherwise, `ready` is sent and the value representing the current state is
+   reported, as above.  Only when the current value reaches a state where
+   `partial` is unset can the initial value be read.
```

Comment:
I find this hard to understand. I would guess if partial is set, you need to continue reading messages until it is unset, then merge them all to get the final value?

Reply:
Yes.  It's a bit complicated to explain this concept in general, but the idea is that the client is supposed to collect (and continuously merge) patches as they come off the wire, but only report the change to the API consumer once `partial` is not set.

In some sense it would make more sense from the viewpoint of the client to have `complete: true` to indicate that we have non-partial data, but in the current normal case we never set `partial` anyway, so we can avoid it entirely, which seemed nice.

Reply:
I've explained this a bit more.

Reply:
Thanks, I find this much clearer now.